### DISCUSSION TITLE
HexagonLayer: Establish bins and max heights before slicing data

### DISCUSSION THREAD

**Discussion by Kirkman** [visgl/deck.gl#6846] [Q&A] [answered]
I'm just dipping my toes in to deck.gl, and was surprised how quickly I was able to make a cool HexagonLayer map of traffic data. 

I'm working on a visualization that shows how the number of traffic accidents has increased over time. My data includes a `year` column, and I managed to successfully add a slider for dynamically adjusting the map to show only the accidents that occurred in specific two-year timeframes. You can see an example here: https://codepen.io/Kirkman/pen/LYeaqVJ

The problem is that the bin positions and the maximum elevations for each of the two-year "timeslices" are different.

Is there a way to precalculate the bin positions, as well as determine the maximum height/elevation across _all_ the years in the data, then provide that when rendering a specific timeslice? That way, each "timeslice" will be using the same scale, and I'll be able to see the number of traffic accidents actually grow as I move the slider from earlier years to later years.

**Comment by Pessimistress** [comment, marked as answer]
To do this, do not change the `data` prop. Always supply your full dataset to the layer. Instead, use a custom `getElevationValue` to determine the elevation of each bin based on the current year. See example in https://deck.gl/docs/api-reference/aggregation-layers/hexagon-layer#elevationaggregation

**Comment by Kirkman** [comment]
I'm running into some obstacles trying to implement @Pessimistress' suggestion. 

The main issue is that if I take the approach of not changing the `data` prop, then the map no longer updates when the `year` input changes. I think this is because the `data` object itself has not changed. What I need to do is force the map to update and re-run the getMax function, which will filter the data after grabbing the updated year from the form. But I can't figure out how to force the map to do that.

You can see where I'm at here: https://codepen.io/Kirkman/pen/rNpbmGp

(FWIW, it also bothers me that this approach forces me to read the value of the `year` input in the DOM every time `getMax()` runs. I'm probably missing something obvious, but I can't find a way to store the year in a place `getMax()` can access without using some sort of global variable. In fact it would make many things easier if there was a way for `getMax` to access _all_ of the binned data, not just the one bin that it's operating on)

**Reply by Pessimistress** [reply]
[updateTriggers](https://deck.gl/docs/api-reference/core/layer#updatetriggers)

**Comment by Kirkman** [comment]
Thanks again, @Pessimistress. I got it working with updateTriggers. 

If anyone else runs into trouble with these same things, I found this link very helpful in understanding updateTriggers: 
https://deck.gl/docs/developer-guide/performance#use-updatetriggers
